control/controlclient/noise.go

```diff
@@ -34,6 +36,77 @@ type noiseConn struct {
 	id   int
 	pool *noiseClient
 	h2cc *http2.ClientConn
+
+	didReadHeaderCh chan struct{} // closed when didReadHeader is set true
+	headerBytesRead uint8
+	headerBuf       [hdrLen]byte
+	readBuf         []byte // consumed, to give back
+	epLen           int
+	epBuf           []byte // accumulate until we can parse it as JSON
+	earlyPayload    *tailcfg.EarlyNoise
+}
+
+func (c *noiseConn) RoundTrip(r *http.Request) (*http.Response, error) {
+	return c.h2cc.RoundTrip(r)
+}
+
+const hdrLen = 9 // http2 frame header size; also size of our early payload size header
+
+func (c *noiseConn) Read(p []byte) (n int, err error) {
+	if need := hdrLen - c.headerBytesRead; need > 0 {
```

Comment:
This is a lot of extra logic in every Read call. Since this is all for a payload that's only seen once at the start of the connection, could you instead do all this at noiseConn construction time, and avoid all this overhead on each future call? As a bonus you can ReadFull and avoid a lot of the fiddly buffer management that you're forced to have inside Read.

Reply:
PTAL